src/gordon_gcp/plugins/event_consumer.py

```diff
@@ -30,15 +30,240 @@
+            self.config['subscription'] = sub
+
+    def _init_auth(self):
+        # a publisher client can't be made with credentials if a channel is
+        # already made/provided, which is what happens when the emulator is
+        # running ಠ_ಠ
+        # See (https://github.com/GodsaogleCloudPlatform/
+        #      google-cloud-python/pull/4839)
+        auth_client = None
+        if not os.environ.get('PUBSUB_EMULATOR_HOST'):
+            scopes = self.config.get('scopes')
+            # creating a dummy `session` as the pubsub.PublisherClient never
+            # uses it but without it aiohttp will complain about an unclosed
+            # client session that would otherwise be made by default
+            auth_client = auth.GAuthClient(
+                keyfile=self.config['keyfile'], scopes=scopes, session='noop')
+        return auth_client
+
+    def _init_subscriber_client(self, auth_client):
+        # Silly emulator constraints
+        creds = getattr(auth_client, 'creds', None)
+        client = pubsub.SubscriberClient(credentials=creds)
+
+        try:
+            client.create_subscription(
+                self.config['subscription'], self.config['topic'])
+
+        except google_exceptions.AlreadyExists:
+            # subscription already exists
+            pass
+
+        except google_exceptions.NotFound as e:
+            msg = f'Topic "{self.config["topic"]}" does not exist.'
+            logging.error(msg, exc_info=e)
+            raise exceptions.GCPGordonError(msg)
+
+        except Exception as e:
+            sub = self.config['subscription']
+            msg = f'Error trying to create subscription "{sub}": {e}'
+            logging.error(msg, exc_info=e)
+            raise exceptions.GCPGordonError(msg)
+
+        return client.subscribe(self.config['subscription'])
+
+    def build_event_consumer(self):
+        self._validate_config()
+        validator = validate.MessageValidator()
+        parser = parse.MessageParser()
+        auth_client = self._init_auth()
+        subscription = self._init_subscriber_client(auth_client)
+        loop = self.kwargs.get('loop', asyncio.get_event_loop())
+
+        return GPSEventConsumer(
+            self.config, subscription, validator, parser, self.success_channel,
+            self.error_channel, loop)
+
+
+class _GPSThread(threading.Thread):
+    """Thread in another asyncio event loop.
+
+    .. attention::
+        This thread class is not meant to be worked with directly.
+
+    Google's Pub/Sub library is blocking. This thread class allows us
+    to work create work in another thread and not block the main event
+    loop.
+
+    Inspiration directly from https://stackoverflow.com/a/29302684
+    """
+    def __init__(self, start_event, name=None, daemon=False):
+        super().__init__(name=name, daemon=daemon)
+        self.loop = None
+        self.thread_id = None
+        self.event = start_event
```

Comment:
This is a `threading.Event` instance, where literally `start` is called on it, so [then the thread knows to start work](https://docs.python.org/3/library/threading.html#event-objects).  The loop hasn't started yet. 

Re: the comment below for docstrings, I realize this whole class could be a bit better documented (even though it's private). I'll add some docs here to make it more obvious.

Reply:
Sorry, I should have been more specific. I was thinking about the `self.event` instance attribute, because it looks like a specific type of event that signals that the thread's event loop has started, and that maybe it could be named to reflect that eg. `self.loop_started`.

Then, in the `run` method, this would read `self.loop.call_soon(self.loop_started.set)` and in the `_thread_pubsub_msg` method it would read `loop_started.wait()`. 

But this is a minor naming thing, so ignore it if it's too nitpicky :)

Reply:
I thought a bit about this, and I think I'm just going to have it as `event`. I think `self.loop_started` is misleading, because below from `_thread_pubsub_msg`, we'd then have: 

```py
thread.start()
loop_started.wait()  # event.wait()
```

I'd be super confused. ¯\\_(ツ)_/¯